nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,631 @@
+    #    block   |  =>     \   /
+    #       \   /           \ /
+    #        \ /             D
+    #         D          forbidden
+
+    #pylint: disable=protected-access
+    if start_node.is_dummy:
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(start_node.node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    else:
+        pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(pred_start_node) != 0 and len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(pred_start_node[0].node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    return attr is None
+
+def check_not_lead_to_duplication_of_activation(sgraph: SearchGraph,
+                                               start_node: SearchGraphNode,
+                                               end_node: SearchGraphNode):
+    """
+    This rule ensures that after the block is deleted there will be no duplication of activation layers.
+    """
+    #         A             A
+    #         |             |
+    #        relu          relu
+    #         |      =>     |
+    #        block         relu
+    #         |
+    #        relu        forbidden
+
+    pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+    next_end_node = sgraph.get_next_nodes(end_node.node_key)
+    if len(next_end_node) == 0 or len(pred_start_node) == 0:
+        return True
+    if pred_start_node[0].is_dummy:
+        pred_start_node = sgraph.get_prev_nodes(pred_start_node[0].node_key)
+
+    if pred_start_node[0].node_type[-1] in PTRELUMetatype.get_all_aliases()\
+         and next_end_node[0].node_type[0] in PTRELUMetatype.get_all_aliases():
+        return False
+    return True
+
+def compare_for_building_block(a: BuildingBlock, b: BuildingBlock):
+    if a.end_node.bottom_id != b.end_node.bottom_id:
+        return a.end_node.bottom_id - b.end_node.bottom_id
+    return b.start_node.main_id - a.start_node.main_id
+
+def check_blocks_combination_is_block(block: BuildingBlock, combination: List[BuildingBlock]) -> bool:
+    """
+    Checks that a combination of blocks is a given block.
+    """
+    if block.start_node.main_id != combination[0].start_node.main_id:
+        return False
+    if block.end_node.bottom_id != combination[-1].end_node.bottom_id:
+        return False
+    i = 0
+    while i < len(combination) - 1:
+        end_i_node = combination[i].end_node
+        start_nexti_node = combination[i + 1].start_node
+        if end_i_node.node_key in start_nexti_node.node_key:
+            i += 1
+            continue
+        return False
+    return True
+
+def search_lin_combination(block: BuildingBlock, blocks: List[BuildingBlock]) -> bool:
+    """
+    Checks that a given block is linear combination of some blocks.
```

Comment:
Are you sure that "a linear combination of blocks" is a conventional mathematical term? If not, you should give a definition or an explanation of just what this means practically.

Reply:
Done.